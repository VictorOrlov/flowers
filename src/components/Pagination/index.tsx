import * as React from "react";

interface IProps {
  current: number;
  total: number;
  perPage: number;
  onPageChange: (page: number) => void;
  pageRange: number;
}

const Pagination: React.SFC<IProps> = (props: IProps) => {
  const pages = () => {
    var pages = [];
    for (var i = rangeStart(); i <= rangeEnd(); i++) {
      pages.push(i);
    }
    return pages;
  };

  const rangeStart = () => {
    var start = props.current - props.pageRange;
    return start > 0 ? start : 1;
  };

  const rangeEnd = () => {
    var end = props.current + props.pageRange;
    return end < props.total ? end : props.total;
  };

  const nextPage = () => {
    return props.current + 1;
  };

  const prevPage = () => {
    return props.current - 1;
  };

  const hasFirst = () => {
    return rangeStart() !== 1;
  };

  const hasLast = () => {
    return rangeEnd() < props.total;
  };

  const hasPrev = () => {
    return props.current > 1;
  };

  const hasNext = () => {
    return props.current < props.total;
  };

  const changePage = (page: any) => {
    props.onPageChange(page);
  };

  return (
    <div className="pagination">
      <div className="pagination__left">
        <span
          className={!hasPrev() ? "hidden_page_numbers" : ""}
          onClick={e => changePage(prevPage())}
        >
          Предыдущая
        </span>
      </div>

      <div className="pagination__mid">
        <ul>
          <li className={!hasFirst() ? "hidden_page_numbers" : ""}>
            <span onClick={e => changePage(1)}>1</span>
          </li>
          <li className={!hasFirst() ? "hidden_page_numbers" : ""}>...</li>
          {pages().map((page, index) => {
            return (
              <li key={index}>
                <span
                  onClick={e => changePage(page)}
                  className={props.current == page ? "current" : ""}
                >
                  {page}
                </span>
              </li>
            );
          })}
          <li className={!hasLast() ? "hidden_page_numbers" : ""}>...</li>
          <li className={!hasLast() ? "hidden_page_numbers" : ""}>
            <span onClick={e => changePage(props.total)}>{props.total}</span>
          </li>
        </ul>
      </div>

      <div className="pagination__right">
        <span
          className={!hasNext() ? "hidden_page_numbers" : ""}
          onClick={e => changePage(nextPage())}
        >
          Следующая
        </span>
      </div>
    </div>
  );
};

Pagination.defaultProps = {
  pageRange: 2
};

export default Pagination;
