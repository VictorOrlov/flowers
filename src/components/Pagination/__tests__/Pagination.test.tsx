import React from 'react';
import { mount } from 'enzyme';
import Pagination from '..';

const mockFn = jest.fn();
const component = mount(
  <Pagination
    current={2}
    total={25}
    perPage={5}
    pageRange={1}
    onPageChange={mockFn}
  />
);

it('snapshot for Pagination Component', ()=> {
  expect(component).toMatchSnapshot();
});

// it('check onClick in Pagination Component', ()=> {
//   console.log(component.find('.pagination__mid').children().props())
//   let prev = component.find(".pagination__left").children();
//   prev.simulate('click');
//   expect(mockFn).toHaveBeenCalledTimes(1)

// })