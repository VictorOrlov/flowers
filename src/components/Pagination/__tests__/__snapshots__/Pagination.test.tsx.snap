// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for Pagination Component 1`] = `
<Pagination
  current={2}
  onPageChange={[MockFunction]}
  pageRange={1}
  perPage={5}
  total={25}
>
  <div
    className="pagination"
  >
    <div
      className="pagination__left"
    >
      <span
        className=""
        onClick={[Function]}
      >
        Предыдущая
      </span>
    </div>
    <div
      className="pagination__mid"
    >
      <ul>
        <li
          className="hidden_page_numbers"
        >
          <span
            onClick={[Function]}
          >
            1
          </span>
        </li>
        <li
          className="hidden_page_numbers"
        >
          ...
        </li>
        <li
          key="0"
        >
          <span
            className=""
            onClick={[Function]}
          >
            1
          </span>
        </li>
        <li
          key="1"
        >
          <span
            className="current"
            onClick={[Function]}
          >
            2
          </span>
        </li>
        <li
          key="2"
        >
          <span
            className=""
            onClick={[Function]}
          >
            3
          </span>
        </li>
        <li
          className=""
        >
          ...
        </li>
        <li
          className=""
        >
          <span
            onClick={[Function]}
          >
            25
          </span>
        </li>
      </ul>
    </div>
    <div
      className="pagination__right"
    >
      <span
        className=""
        onClick={[Function]}
      >
        Следующая
      </span>
    </div>
  </div>
</Pagination>
`;
